lib/veri.ml

```diff
@@ -148,41 +149,49 @@ class ['a] t arch dis =
     method private update_event ev =
       SM.update (fun c -> c#register_event ev)
 
-    (** [resolve_var var] - returns a result, bound with [var].
-        Sequence of searches is the following:
-        1) among read events that occured at current step in the same context,
-           with the same variable;
-        2) among read events that occures at current step, in other context,
-           with the same variable;
-        3) in current context, for the same variable *)
-    method private resolve_var : var -> 'a r = fun var ->
+    (** [touch x] - return a value bound with [x] where
+        [x] is either address or variable. In each variant of x
+        an appropriative lookup order is applied.
+        An event could be emited. *)
+    method private touch x =
+      let touch_value probes =
+        List.find_map probes ~f:(fun (probe, x) -> match probe () with
+            | None -> None
+            | Some mv -> Some (mv, x)) |> function
+        | None -> SM.return None
+        | Some (mv, `Lookup) -> SM.return (Some (Move.data mv))
+        | Some (mv, `Emit f) ->
+          self#update_event (f mv) >>= fun () ->
+          SM.return (Some (Move.data mv)) in
       SM.get () >>= fun ctxt ->
-      match find_reg_read (self_events ctxt) (same_var var) with
-      | Some mv -> self#eval_exp (Bil.int (Move.data mv))
-      | None ->
-        match find_reg_read (other_events ctxt) (same_var var) with
-        | Some mv -> self#eval_exp (Bil.int (Move.data mv))
-        | None -> super#lookup var
-
-    (** [lookup var] - returns a result, bound with variable.
-        Search starts from self events, if it was write access to given
-        variable at current step. And if it was, then result of write
-        access returned and no read event is emitted.
-        Otherwise searching continues as written above for [resolve_var],
-        with emitting register_read event. *)
+      let self_evs = self_events ctxt in
+      let other_evs = other_events ctxt in
+      let mk_move f mv = f (Move.cell mv) (Move.data mv) in
+      match x with
+      | `Var var ->
```

Comment:
These two branches looks like Siamese, you're still missing an abstraction here.